chart.js.d.ts

```diff
@@ -0,0 +1,605 @@
+    datasetIndex: number
+  }
+
+  interface ChartTooltipsOptions {
+    enabled?: boolean
+    custom?: (tooltip: Tooltip) => void // Check required, see Tooltip interface
+    mode?: string
+    backgroundColor?: string
+    titleFontFamily?: string
+    titleFontSize?: number
+    titleFontStyle?: string
+    titleSpacing?: number
+    titleColor?: string
+    titleAlign?: string
+    bodyFontFamily?: string
+    bodyFontSize?: number
+    bodyFontStyle?: string
+    bodySpacing?: number
+    bodyColor?: string
+    bodyAlign?: string
+    footerFontFamily?: string
+    footerFontSize?: number
+    footerFontStyle?: string
+    footerSpacing?: number
+    footerColor?: string
+    footerAlign?: string
+    titleMarginBottom?: number
+    footerMarginTop?: number
+    xPadding?: number
+    yPadding?: number
+    caretSize?: number
+    cornerRadius?: number
+    multiKeyBackground?: string
+    callbacks?: {
+      // @formatter:off
+      // Checks required, correct args? title and label specify own args in docs, which are required?
+      beforeTitle?:   (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      title?:         (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      afterTitle?:    (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      beforeBody?:    (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      beforeLabel?:   (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      label?:         (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      afterLabel?:    (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      afterBody?:     (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      beforeFooter?:  (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      footer?:        (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      afterFooter?:   (xLabel: string | Array<string>, yLabel: string | Array<string>, index: number, data: Object) => string | Array<string>
+      // @formatter:on
+    }
+  }
+
+  // What fields does a tooltip object passed to custom contain?
+  interface Tooltip {
+
+  }
+
+  interface ChartAnimationOptions {
+    duration?: number
+    easing?: string
+    onProgress?: (args: {chart: ChartInstance, animationDetails: AnimationDetails}) => void // Check required, what fields should animationDetails have?
```

Comment:
No required. Allows external synchronization with animations. See https://github.com/nnnick/Chart.js/blob/master/samples/AnimationCallbacks/progress-bar.html

### Animation detail
```javascript
animationDetail = {
  chartInstance: // chart object
  animationObject: // animation object
}
```

###  Animation object
This is of type `Chart.Animation`
```javascript
{
  currentStep: // number or null - current step of the animation.
  numSteps: // number - frame count for the animation
  easing: // string - Chart easing. See allowed easings below
  render: // function - callback that renders the chart for the frame
  onAnimationProgress: // function - user defined callback
  onAnimationComplete: // function - user defined callback
}
```

### Easings
The following easings are allowed

- linear
- easeInQuad
- easeOutQuad
- easeInOutQuad
- easeInCubic
- easeOutCubic
- easeInOutCubic
- easeInQuart
- easeOutQuart
- easeInOutQuart
- easeInQuint
- easeOutQuint
- easeInOutQuint
- easeInSine
- easeOutSine
- easeInOutSine
- easeInExpo
- easeOutExpo
- easeInOutExpo
- easeInCirc
- easeOutCirc
- easeInOutCirc
- easeInElastic
- easeOutElastic
- easeInOutElastic
- easeInBack
- easeOutBack
- easeInOutBack
- easeInBounce
- easeOutBounce
- easeInOutBounce